google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,936 @@
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutUpdate).Minutes())
+
+	cluster := &dataproc.Cluster{
+		ClusterName: clusterName,
+		ProjectId:   project,
+		Config:      &dataproc.ClusterConfig{},
+	}
+
+	updMask := []string{}
+
+	if d.HasChange("labels") {
+		v := d.Get("labels")
+		m := make(map[string]string)
+		for k, val := range v.(map[string]interface{}) {
+			m[k] = val.(string)
+		}
+		cluster.Labels = m
+
+		updMask = append(updMask, "labels")
+	}
+
+	if d.HasChange("cluster_config.0.worker_config.0.num_instances") {
+		wconfigs := d.Get("cluster_config.0.worker_config").([]interface{})
+		conf := wconfigs[0].(map[string]interface{})
+
+		desiredNumWorks := conf["num_instances"].(int)
+		cluster.Config.WorkerConfig = &dataproc.InstanceGroupConfig{
+			NumInstances: int64(desiredNumWorks),
+		}
+
+		updMask = append(updMask, "config.worker_config.num_instances")
+	}
+
+	if d.HasChange("cluster_config.0.preemptible_worker_config.0.num_instances") {
+		wconfigs := d.Get("cluster_config.0.preemptible_worker_config").([]interface{})
+		conf := wconfigs[0].(map[string]interface{})
+
+		desiredNumWorks := conf["num_instances"].(int)
+		cluster.Config.SecondaryWorkerConfig = &dataproc.InstanceGroupConfig{
+			NumInstances: int64(desiredNumWorks),
+		}
+
+		updMask = append(updMask, "config.secondary_worker_config.num_instances")
+	}
+
+	if len(updMask) > 0 {
+		patch := config.clientDataproc.Projects.Regions.Clusters.Patch(
+			project, region, clusterName, cluster)
+		op, err := patch.UpdateMask(strings.Join(updMask, ",")).Do()
+		if err != nil {
+			return err
+		}
+
+		// Wait until it's updated
+		waitErr := dataprocClusterOperationWait(config, op, "updating Dataproc cluster ", timeoutInMinutes, 2)
+		if waitErr != nil {
+			return waitErr
+		}
+
+		log.Printf("[INFO] Dataproc cluster %s has been updated ", d.Id())
+	}
+
+	return resourceDataprocClusterRead(d, meta)
+}
+
+func resourceDataprocClusterRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region := d.Get("region").(string)
+	clusterName := d.Get("name").(string)
+
+	cluster, err := config.clientDataproc.Projects.Regions.Clusters.Get(
+		project, region, clusterName).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataproc Cluster %q", clusterName))
+	}
+
+	d.Set("name", cluster.ClusterName)
+	d.Set("region", region)
+	d.Set("labels", cluster.Labels)
+
+	cfg, err := flattenClusterConfig(d, cluster.Config)
+	if err != nil {
+		return err
+	}
+
+	d.Set("cluster_config", cfg)
+	return nil
+}
+
+func flattenClusterConfig(d *schema.ResourceData, cfg *dataproc.ClusterConfig) ([]map[string]interface{}, error) {
+
+	data := getOrCreateNewMapSchema(d, "cluster_config")
```

Comment:
I'm trying to figure out what this is used for. Is it because some fields aren't returned from the API? If so, the way we usually handle that is to do a d.Get() for those fields specifically. This makes me a bit nervous that we might at some point forget to populate a new field that gets added and not notice because it just kept whatever was already in state.